htmresearch/frameworks/nlp/imbu.py

```diff
@@ -0,0 +1,203 @@
+import argparse
+import os
+
+from enum import Enum
+
+from htmresearch.encoders import EncoderTypes
+from htmresearch.frameworks.nlp.classification_model import ClassificationModel
+from htmresearch.frameworks.nlp.classify_fingerprint import (
+  ClassificationModelFingerprint)
+from htmresearch.frameworks.nlp.classify_htm import ClassificationModelHTM
+from htmresearch.frameworks.nlp.classify_keywords import (
+  ClassificationModelKeywords)
+from htmresearch.frameworks.nlp.classify_windows import (
+  ClassificationModelWindows)
+from htmresearch.support.csv_helper import readCSV
+from htmresearch.support.text_preprocess import TextPreprocess
+
+
+
+class ImbuModelTypes(Enum):
+  CioWordFingerprint = ClassificationModelFingerprint
+  CioDocumentFingerprint = ClassificationModelFingerprint
+  CioWindows = ClassificationModelWindows
+  Keywords = ClassificationModelKeywords
+  HTMNetwork = ClassificationModelHTM
+
+
+
+class ModelSimilarityMetrics(Enum):
+  pctOverlapOfInput = "pctOverlapOfInput"
+
+
+
+class ImbuError(Exception):
+  pass
+
+
+
+class ImbuUnableToLoadModelError(ImbuError):
+  def __init__(self, exc):
+    self.exc = exc
+
+
+
+class ImbuModels(object):
```

Comment:
Intent is to unify imbu_runner.py and fluent_api.py with a formal and stable API.  The Imbu web app has not been stable due to frequent changes in `nupic.research`.  Furthermore, the web app is made a bit more complicated than it needs to be.  It has some infrastructure in place to support background parallel training, which required an ad-hoc RPC mechanism that required some duplication of the API of the ClassificationModel in the web app itself -- a maintenance headache.  We'll be removing the RPC mechanism and duplication from the web app in favor of pre-training the models externally with this tool, which also defines the API for consumption by the web app.

See IMBU-70 for background.